populator-machinery/controller_test.go

```diff
@@ -0,0 +1,538 @@
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+			},
+			expectedResult: nil,
+			expectedKeys:   []string{storageClassKey},
+		},
+		{
+			name:         "PVC not bound to a node",
+			key:          "pvc/" + testPvcNamespace + "/" + testPvcName,
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, "", testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+			},
+			expectedResult: nil,
+			expectedKeys:   []string{},
+		},
+		{
+			name:         "Create populator pod",
+			key:          "pvc/" + testPvcNamespace + "/" + testPvcName,
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+			},
+			expectedResult: nil,
+			expectedKeys:   []string{podKey, pvcPrimeKey},
+		},
+		{
+			name:         "Wait populator pod succeed",
+			key:          "pvc/" + testPvcNamespace + "/" + testPvcName,
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+				pod(corev1.PodRunning),
+			},
+			expectedResult: nil,
+			expectedKeys:   []string{podKey, pvcPrimeKey},
+		},
+		{
+			name:         "Populator pod failed",
+			key:          "pvc/" + testPvcNamespace + "/" + testPvcName,
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+				pod(corev1.PodFailed),
+			},
+			expectedResult: nil,
+			expectedKeys:   []string{podKey, pvcPrimeKey},
+		},
+
+		{
+			name:         "Data populate succeeded, pvcPrime not exists",
+			key:          "pvc/" + testPvcNamespace + "/" + testPvcName,
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+				pod(corev1.PodSucceeded),
+			},
+			expectedResult: errors.New("Failed to find PVC for populator pod"),
```

Comment:
comment for future PRs: We should capture the Pod and PVC name in the error log.

Reply:
Good point! I want to do this update with a separate PR, and keep this just for adding tests